astrolabe/runner.py

```diff
@@ -163,34 +167,72 @@ def run(self, persist_cluster=False, startup_time=1):
             driver_workload=self.spec.driverWorkload,
             startup_time=startup_time)
 
-        # Step-3: begin maintenance routine.
-        final_config = self.spec.maintenancePlan.final
-        cluster_config = final_config.clusterConfiguration
-        process_args = final_config.processArgs
-
-        if not cluster_config and not process_args:
-            raise RuntimeError("invalid maintenance plan")
-
-        if cluster_config:
-            LOGGER.info("Pushing cluster configuration update")
-            self.cluster_url.patch(**cluster_config)
-
-        if process_args:
-            LOGGER.info("Pushing process arguments update")
-            self.cluster_url.processArgs.patch(**process_args)
-
-        # Sleep before polling to give Atlas time to update cluster.stateName.
-        sleep(3)
-
-        # Step-4: wait until maintenance completes (cluster is IDLE).
-        selector = BooleanCallablePoller(
-            frequency=self.config.polling_frequency,
-            timeout=self.config.polling_timeout)
-        LOGGER.info("Waiting for cluster maintenance to complete")
-        selector.poll([self], attribute="is_cluster_state", args=("IDLE",),
-                      kwargs={})
-        self.verify_cluster_configuration_matches("final")
-        LOGGER.info("Cluster maintenance complete")
+        for operation in self.spec.operations:
+            if hasattr(operation, 'setClusterConfiguration'):
+                # Step-3: begin maintenance routine.
+                final_config = operation.setClusterConfiguration
+                cluster_config = final_config.clusterConfiguration
+                process_args = final_config.processArgs
+
+                if not cluster_config and not process_args:
+                    raise RuntimeError("invalid maintenance plan")
+
+                if cluster_config:
+                    LOGGER.info("Pushing cluster configuration update")
+                    self.cluster_url.patch(**cluster_config)
+
+                if process_args:
+                    LOGGER.info("Pushing process arguments update")
+                    self.cluster_url.processArgs.patch(**process_args)
+
+                # Sleep before polling to give Atlas time to update cluster.stateName.
+                sleep(3)
```

Comment:
I think instead of waiting here, we should introduce a wait at the start of `wait_for_idle()`. That will provide a better abstraction and will prevent these somewhat random sleep times from being scattered all over the codebase.

Reply:
Actually this sleep isn't necessary anymore because the wait is now part of the scenarios since 3 seconds was not always sufficient. I'll remove it.